src/processor.rs

```diff
@@ -1,18 +1,171 @@
-    _program_id: &Pubkey,
-    _accounts: &[AccountInfo],
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
 ) -> ProgramResult {
+    let accounts_iter = &mut accounts.iter();
+
+    let holder_rewards_pool_info = next_account_info(accounts_iter)?;
+    let extra_metas_info = next_account_info(accounts_iter)?;
+    let mint_info = next_account_info(accounts_iter)?;
+    let mint_authority_info = next_account_info(accounts_iter)?;
+    let _system_program_info = next_account_info(accounts_iter)?;
+
+    // Run checks on the mint.
+    {
+        // Ensure the mint is owned by SPL Token-2022.
+        if !mint_info.owner.eq(&spl_token_2022::id()) {
+            return Err(ProgramError::InvalidAccountOwner);
+        }
+
+        let mint_data = mint_info.try_borrow_data()?;
+        let mint = StateWithExtensions::<Mint>::unpack(&mint_data)?;
+
+        // Ensure the mint is configured with the `TransferHook` extension,
+        // and the program ID is the Paladin Rewards program.
+        let transfer_hook = mint.get_extension::<TransferHook>()?;
+        let hook_program_id: Option<Pubkey> = transfer_hook.program_id.into();
+        if !hook_program_id.eq(&Some(*program_id)) {
+            return Err(PaladinRewardsError::IncorrectTransferHookProgramId.into());
+        }
+
+        // Ensure the provided mint authority is the correct mint authority.
+        if !mint
+            .base
+            .mint_authority
+            .eq(&COption::Some(*mint_authority_info.key))
+        {
+            return Err(PaladinRewardsError::IncorrectMintAuthority.into());
+        }
+
+        // Ensure the mint authority is a signer.
+        if !mint_authority_info.is_signer {
+            return Err(ProgramError::MissingRequiredSignature);
+        }
+    }
+
+    // Initialize the holder rewards pool account.
+    {
+        let (holder_rewards_pool_address, holder_rewards_pool_bump) =
+            get_holder_rewards_pool_address_and_bump_seed(mint_info.key);
+        let holder_rewards_pool_signer_seeds = &[
+            SEED_PREFIX_HOLDER_REWARDS_POOL,
+            mint_info.key.as_ref(),
+            &[holder_rewards_pool_bump],
+        ];
+
+        // Ensure the provided holder rewards pool address is the correct
+        // address derived from the mint.
+        if !holder_rewards_pool_info
+            .key
+            .eq(&holder_rewards_pool_address)
+        {
+            return Err(PaladinRewardsError::IncorrectHolderRewardsPoolAddress.into());
+        }
+
+        // Ensure the holder rewards pool account has not already been
+        // initialized.
+        if holder_rewards_pool_info.data_len() != 0 {
+            return Err(ProgramError::AccountAlreadyInitialized);
+        }
+
+        // Allocate & assign.
+        invoke_signed(
+            &system_instruction::allocate(
+                &holder_rewards_pool_address,
+                std::mem::size_of::<HolderRewardsPool>() as u64,
```

Comment:
Seeing this brings up the question -- is there any plan to add account discriminants? I don't have a strong opinion, except for the general "it's better to have a disciminant to avoid the issues that token-2022 had"

Reply:
I'm leaning towards not using discriminators here. If we go with single-byte discriminators, then it will throw off the alignment for both accounts. If we use 8 bytes, that means we're increasing holder rewards state by 50% and holder rewards pool state by 100%.

I think considering these are quite small, a length check is more appropriate. What do you think?

Reply:
Yeah ok, works for me!